### PR TITLE
Fix issues of GPS metadata handling

### PR DESCRIPTION
**Pull request by nmbgeek**
#2541 #2635 #2616 Directly address these issues

#2260 Address some of the issues in the comments

This pull request adds handling null GPS coordinates and also coordinates that are not a string.

### CODE REVIEW

**Comment by nmbgeek** [comment]
I believe this should also fix #2616 because $rawData can be an array.

File: lib/Listener/ExifMetadataProvider.php [hunk 1]
```diff
@@ -109,11 +109,15 @@ public function handle(Event $event): void {
 	 * @param array|string $coordinates
 	 */
 	private function gpsDegreesToDecimal($coordinates, ?string $hemisphere): float {
+		if (is_null($coordinates)) {
+			throw new \Exception('Coordinates are null');
+		}
+
 		if (is_string($coordinates)) {
-			$coordinates = array_map('trim', explode(',', $coordinates));
+			$coordinates = array_map("trim", explode(",", $coordinates));
```
Comment (artonge) [inline lib/Listener/ExifMetadataProvider.php:117]: Our coding standard asks for the original :)

File: lib/Listener/ExifMetadataProvider.php [hunk 2]
```diff
@@ -123,7 +127,15 @@ private function gpsDegreesToDecimal($coordinates, ?string $hemisphere): float {
 		return $sign * ($degrees + $minutes / 60 + $seconds / 3600);
 	}
 
-	private function parseGPSData(string $rawData): float {
+	private function parseGPSData($rawData): float {
```
Comment (artonge) [inline lib/Listener/ExifMetadataProvider.php:130]:
```suggestion
	/**
	 * @param array|string $rawData
	 */
	private function parseGPSData($rawData): float {
```

File: lib/Listener/ExifMetadataProvider.php [hunk 3]
```diff
@@ -123,7 +127,15 @@ private function gpsDegreesToDecimal($coordinates, ?string $hemisphere): float {
...
+		if (is_array($rawData)) {
+		if (count($rawData) === 3) {
+			return floatval($rawData[0]) + floatval($rawData[1]) / 60 + floatval($rawData[2]) / 3600;
+			} else {
+				return 0;  // Handle unexpected array structure
+			}
+		}
+
```
Comment (artonge) [inline lib/Listener/ExifMetadataProvider.php:138]: Weird indentation.

**Review by artonge — COMMENTED** [review]
Looks good overall. Minor comment on code styling.
Thanks for looking into it :heart:

File: lib/Listener/ExifMetadataProvider.php (hunk 3 above)
Comment (nmbgeek) [reply lib/Listener/ExifMetadataProvider.php]: I think these are all fixed now.

File: lib/Listener/ExifMetadataProvider.php [hunk 4]
```diff
@@ -122,8 +126,19 @@ private function gpsDegreesToDecimal($coordinates, ?string $hemisphere): float {
...
-
-	private function parseGPSData(string $rawData): float {
+	
+	/**
+	 * @param array|string $rawData
+	 */
+	private function parseGPSData($rawData): float {
+		if (is_array($rawData)) {
+			if (count($rawData) === 3) {
+				return floatval($rawData[0]) + floatval($rawData[1]) / 60 + floatval($rawData[2]) / 3600;
```
Comment (artonge) [inline lib/Listener/ExifMetadataProvider.php:136]: Why are we doing that computation here? It looks like the coordinate computation that we are doing above.
Could you add some comment explaining the different formats?

Comment (nmbgeek) [reply lib/Listener/ExifMetadataProvider.php:136]: After reviewing the correct implementation of EXIF metadata for altitude you are correct, there is no need for this.  Apparently some cameras/applications add invalid metadata.  Altitude in only parsed with this parseGPSData and I was trying to correct for that.  I believe just adding a string check to line 98 may be the better solution?

`if (array_key_exists('GPSAltitude', $rawExifData['GPS']) && array_key_exists('GPSAltitudeRef', $rawExifData['GPS']) && is_string($rawExifData['GPS']['GPSAltitude'])) {`

Comment (artonge) [reply lib/Listener/ExifMetadataProvider.php]: If this only occur with `GPSAltitude`, then yes. But I am wondering if we could handle the array format. Maybe add the `if` condition to prevent the error, but also log the value so we can have more information to improve the handling in the future?

Comment (nmbgeek) [reply lib/Listener/ExifMetadataProvider.php]: #2541 Has an image that has Altitude as '0 0 0' and breaks without checking if it is a string.  The GPS Data also appears to have been scrubbed from the image instead of just removing it with latitude and longitude also both being 0.  From reading some exif documentation it should just be a rational64u formatted as numerator/denominator.  If the data isn't valid is there really anything else to do with it?  It doesn't appear that Photos is actually doing anything with the altitude value either.

File: lib/Listener/ExifMetadataProvider.php [hunk 5]
```diff
@@ -95,9 +95,9 @@ public function handle(Event $event): void {
 				$gps['longitude'] = $this->gpsDegreesToDecimal($rawExifData['GPS']['GPSLongitude'], $rawExifData['GPS']['GPSLongitudeRef']);
 			}
 
-			if (array_key_exists('GPSAltitude', $rawExifData['GPS']) && array_key_exists('GPSAltitudeRef', $rawExifData['GPS'])) {
-				$gps['altitude'] = ($rawExifData['GPS']['GPSAltitudeRef'] === "\u{0000}" ? 1 : -1) * $this->parseGPSData($rawExifData['GPS']['GPSAltitude']);
-			}
+			if (array_key_exists('GPSAltitude', $rawExifData['GPS']) && array_key_exists('GPSAltitudeRef', $rawExifData['GPS']) && is_string($rawExifData['GPS']['GPSAltitude'])) {
+				$gps['altitude'] = ($rawExifData['GPS']['GPSAltitudeRef'] === '1' ? -1 : 1) * $this->parseGPSData($rawExifData['GPS']['GPSAltitude']);
```
Comment (artonge) [inline lib/Listener/ExifMetadataProvider.php:99]: Is this change necessary ?

File: lib/Listener/ExifMetadataProvider.php [hunk 6]
```diff
@@ -95,9 +95,9 @@ public function handle(Event $event): void {
...
+			}
```
Comment (artonge) [inline lib/Listener/ExifMetadataProvider.php:100]: Extra white space

File: lib/Listener/ExifMetadataProvider.php (hunk 5 above)
Comment (nmbgeek) [reply lib/Listener/ExifMetadataProvider.php:99]: This was originally `$rawExifData['GPS']['GPSAltitudeRef'] === "\u{0000}" ? 1 : -1)` \u{0000} is supposed to be the null character but the specifications say it can also be 0 for above sea level. A value of 1 indicates below sea level.

Comment (artonge) [reply lib/Listener/ExifMetadataProvider.php:99]: Hmm, alright then. Have you tested it?

Comment (nmbgeek) [reply lib/Listener/ExifMetadataProvider.php:99]: Yes. I tested with this code and expected values were logged.


```
			if (array_key_exists('GPSAltitude', $rawExifData['GPS']) && array_key_exists('GPSAltitudeRef', $rawExifData['GPS']) && is_string($rawExifData['GPS']['GPSAltitude'])) {
				$this->logger->info('GPSAltitude: ' . $rawExifData['GPS']['GPSAltitude']);
				$this->logger->info('GPSAltitudeRef: ' . $rawExifData['GPS']['GPSAltitudeRef']);
				$gps['altitude'] = ($rawExifData['GPS']['GPSAltitudeRef'] === '1' ? -1 : 1) * $this->parseGPSData($rawExifData['GPS']['GPSAltitude']);
				$this->logger->info('GPSAltitudeCalc: ' . $gps['altitude']);
			}
```

Comment (nmbgeek) [reply lib/Listener/ExifMetadataProvider.php:99]: I tested with a photo that I used exiftool to write metadata for below sea level and it didn't work.  The logged value for GPSAltitudeRef was "\u0001" so I will revert it to using the unicode character but assuming above sea level if the value isn't explicitly below.

**Review by artonge — APPROVED** [review]

**Comment by artonge** [comment]
@nmbgeek can you squash your changes?

**Comment by artonge** [comment]
My bad, I rebased your branch to run the CI against the latest master. This should help you:

```shell
git reset --hard HEAD~11
git cherry-pick b674472a58b9fb2535374f3cb4e72e6155e9c563
git push --force-with-lease
```
